pkg/eraser/eraser.go

```diff
@@ -0,0 +1,248 @@
+	if protocol, addr, err = parseEndpoint(endpoint); err != nil && protocol == "" {
+		fallbackEndpoint := fallbackProtocol + "://" + endpoint
+		protocol, addr, err = parseEndpoint(fallbackEndpoint)
+	}
+	return
+}
+
+func parseEndpoint(endpoint string) (string, string, error) {
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return "", "", err
+	}
+
+	switch u.Scheme {
+	case "tcp":
+		return "tcp", u.Host, nil
+
+	case "unix":
+		return "unix", u.Path, nil
+
+	case "":
+		return "", "", fmt.Errorf("using %q as endpoint is deprecated, please consider using full url format", endpoint)
+
+	default:
+		return u.Scheme, "", fmt.Errorf("protocol %q not supported", u.Scheme)
+	}
+}
+
+func getImageClient(context *cli.Context) (pb.ImageServiceClient, *grpc.ClientConn, error) {
+	addr, dialer, err := GetAddressAndDialer("unix:///run/containerd/containerd.sock")
+
+	if err != nil {
+		fmt.Print("get address and dialer err")
+	}
+
+	conn, err := grpc.Dial(addr, grpc.WithInsecure(), grpc.WithBlock(), grpc.WithTimeout(Timeout), grpc.WithContextDialer(dialer))
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	imageClient := pb.NewImageServiceClient(conn)
+
+	return imageClient, conn, nil
+}
+
+func ListImages(client pb.ImageServiceClient, image string) (resp *pb.ListImagesResponse, err error) {
+	request := &pb.ListImagesRequest{Filter: &pb.ImageFilter{Image: &pb.ImageSpec{Image: image}}}
+
+	resp, err = client.ListImages(context.Background(), request)
+
+	return
+}
+
+func RemoveImage(client pb.ImageServiceClient, image string) (resp *pb.RemoveImageResponse, err error) {
+	if image == "" {
+		return nil, fmt.Errorf("ImageID cannot be empty")
+	}
+
+	request := &pb.RemoveImageRequest{Image: &pb.ImageSpec{Image: image}}
+
+	resp, err = client.RemoveImage(context.Background(), request)
+
+	return
+}
+
+func contains(slice []string, str string) bool {
+	for _, s := range slice {
+		if s == str {
+			return true
+		}
+	}
+	return false
+}
+
+func removeDuplicateValues(intSlice []string) []string {
+	keys := make(map[string]bool)
+	list := []string{}
+
+	for _, entry := range intSlice {
+		if _, value := keys[entry]; !value {
+			keys[entry] = true
+			list = append(list, entry)
+		}
+	}
+	return list
+}
+
+func main() {
+
+	// image job can pass the imagelist into each pod as a env variable maybe
+
+	ctx := cli.NewContext(nil, nil, nil)
+
+	imageClient, conn, err := getImageClient(ctx)
+
+	if err != nil {
+		fmt.Printf("image client err")
+	}
+
+	r, err := ListImages(imageClient, "")
+	if err != nil {
+		fmt.Printf("list err")
+	}
+
+	var allImages []string
+	m := make(map[string][]string)
+
+	for _, img := range r.Images {
+		allImages = append(allImages, img.Id)
+		m[img.Id] = img.RepoTags
+	}
+
+	response, err := pb.NewRuntimeServiceClient(conn).ListContainers(context.Background(), new(pb.ListContainersRequest))
+
+	if err != nil {
+		fmt.Printf("list containers err")
+	}
+
+	var runningImages []string
```

Comment:
You can make this a `map[string]struct{}` and it will automatically de-up as you insert and provides an index so you don't need to iterate the list to determine if an image is in the list.

Reply:
what would the struct be?

Reply:
In this case we are just using the map for a lookup value, so this would either be an empty struct (`struct{}`) where we throw away the value. Example: https://play.golang.org/p/DI1I08rZRom

So we'd use this like so:

```go
for _, img := range allImages {
    if _, isRunning := runningImages[img]; !isRunning {
        nonRunningImages = append(nonRunningImages, img)
    }
}
```

We are using the map to tell us if the image is in the running list rather than iterating over all the objects in `runningImages`

Reply:
got it thank you!